File: homework/quang_tu/path-sum.cpp
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, 0, targetSum);
    }

    bool dfs(TreeNode* root, int sum, int target) {
        if (!root)
            return false;

        bool result = false;

        sum += root->val;

        if (!root->left && !root->right)
            result = sum == target;

        return result || dfs(root->left, sum, target) ||
            dfs(root->right, sum, target);
    }
};

File: homework/Luong Viet Hai/btvn-7/PathSum.php
<?php

class TreeNode
{
    public $val = null;
    public $left = null;
    public $right = null;

    function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class Solution
{

    /**
     * @param TreeNode $root
     * @param Integer $targetSum
     * @return Boolean
     */

    protected $sum;
    protected $isSuccess = false;

    function hasPathSum($root, $targetSum)
    {
        $this->sum = $targetSum;
        $this->dfp($root, 0);

        return $this->isSuccess;
    }

    public function dfp(?TreeNode $node, $parentVal)
    {
        if ($node == null) return 0;
        if ($parentVal + $node->val === $this->sum and $node->left == null and $node->right == null) {
            $this->isSuccess = true;
            return $parentVal + $node->val;
        }

        $this->dfp($node->right, $parentVal + $node->val);
        $this->dfp($node->left, $parentVal + $node->val);

        return $parentVal + $node->val;
    }
}

$testNode = new TreeNode(1, (new TreeNode(2, null, null)), null);

$solution = new Solution();

var_dump($solution->hasPathSum($testNode, 1));